install

```diff
@@ -602,6 +602,8 @@ case "${dist}-${code}" in
         code="bionic"
         is_ubuntu_dist=true
         is_ubuntu_bionic=true
+        # dirty hack for bionic since postgresql has the wrong locale otherwise
+        localectl set-locale LANG=en_US.utf8
```

Comment:
unfortunately this self described dirty hack is blocking my attempts to build a single container docker image. initially the error message is that `localectl` can't be found, but even if i install systemd onto the minimal docker bionic image, i still get a `Failed to create bus connection: No such file or directory` error that i think gets deep into the nature of the docker beast. what does postgresql do without this fix? how can i check whether alternative dirty hacks solve the problem?

i did notice these lines in the libretime-core dockerfile within David's multicontainer build:

        locale-gen "en_US.UTF-8"
        echo -e "LC_ALL=en_US.UTF-8\nLANG=en_US.UTF-8" >> /etc/default/locale
        [...]
        ENV LC_ALL=en_US.UTF-8
        ENV LANG=en_US.UTF-8
        ENV LANGUAGE=en_US.UTF-8

Reply:
Without the hack PostgreSQL doesn't get configured with the proper locale and the setup routine later fails when it tries to create the database. IMO this is part of how the Debian PostgreSQL package is trying to do an opinionated
config of PostgreSQL.

Personally I don't think the the monolithic docker image approach is how docker is intended to be used. Possibly alternative containerizations technologies might fit the bill much nicer if you really need to run systemd and all it's bits and pieces inside a container.

Could this be solved by using packages inside of the image? IMO the whole install script has always been a dirty hack written by legacy upstream to help some users test the latest code. As far as I can tell it has done about the same amount of harm as it has done good (ie. @radiorabe it killed one of our developers Linux install because he didn't run it in a VM when we just started out with legacy upstream).

Reply:
Can `locale-gen` not be used instead and avoid the dependency on systemd (and the message bus)? That way container solutions that don't provide a message bus, such as Docker (I think that is the issue), can still be support

Reply:
I'll give it a try... Would `locale-gen` alone work or does it also need to touch `/etc/default/locale`?

Reply:
I'm also not sure how the locale hack would behave in a non Vagrant context given that we have [this](https://github.com/LibreTime/libretime/releases#issues-no-locale-3.0.0-alpha.6) in the release notes so any suggestions to fix it are very welcome.

I isolated it in the bionic switch to make sure that old support stuff doesn't break for this reason.

Could it make sense to address the locale stuff in another PR so we can get PHP 7.2 out of the door? I'd probably also like to give `locale-gen "C"` a try for that matter.

Reply:
reporting back here until another locale fix PR is created. simply replacing 

         localectl set-locale LANG=en_US.utf8

with 

        locale-gen "en_US.UTF-8"
        echo -e "LC_ALL=en_US.UTF-8\nLANG=en_US.UTF-8" >> /etc/default/locale

allows the install process to continue, but does not solve the postgresql issue. errors in the `install` report log:

```
Setting up postgresql-10 (10.6-0ubuntu0.18.04.1) ...
/var/lib/dpkg/info/postgresql-10.postinst: 2: /etc/default/locale: -e: not found
dpkg: error processing package postgresql-10 (--configure):
 installed postgresql-10 package post-installation script subprocess returned error exit status 127

[...]

dpkg: dependency problems prevent configuration of postgresql:
 postgresql depends on postgresql-10; however:
  Package postgresql-10 is not configured yet.

dpkg: error processing package postgresql (--configure):
 dependency problems - leaving unconfigured

[...]

Errors were encountered while processing:
 postgresql-10
 postgresql
E: Sub-process /usr/bin/dpkg returned an error code (1)

[...]

-----------------------------------------------------
              * Configuring PostgreSQL *              
-----------------------------------------------------
System has not been booted with systemd as init system (PID 1). Can't operate.
```

Reply:
> on the contrary, my understanding is that it's best to try to keep layers small.

See [here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers) and [here](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run) for best practices pertaining to layer sizing in docker. There ist also some info on [decoupling applications](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#decouple-applications). I understand this last link as a recommendation to not use an init system in a container unless you can't find a way around it:

> Limiting each container to one process is a good rule of thumb, but it is not a hard and fast rule.

In LibreTime's case there is nothing that really depends on something taking over PID 1 duties except for the installer. That is unless you want to do a bloated "full" LibreTime container which I personally wouldn't recommend and thus can't help you out with.

```
-----------------------------------------------------
              * Configuring PostgreSQL *              
-----------------------------------------------------
System has not been booted with systemd as init system (PID 1). Can't operate.
```

You seem to run into this during `docker build`. Since docker governs the build process it doesn't run it with systemd as PID 1 and thus the automated postgresql install can't work. You can get around this by calling the installer with `--no-postgres` so it skips trying to install and configure PostgreSQL but then you would need to take care of PostgreSQL beforehand. 

There are some examples on how to install PostgreSQL in the official library/postgres images at https://github.com/docker-library/postgres but they all assume that the PostgreSQL processes get to run in their own containers.

Reply:
thanks for all your help. i definitely have no objection to a multi-container setup. in fact that's what we're using at my station. my goals with the docker build are simplicity/best practices, and also staying as close to upstream as possible, ie. avoiding fixes like https://github.com/ned-kelly/docker-multicontainer-libretime/issues/13

i'm going to continue this conversation in other threads.

Reply:
Resolving this since I removed the locale hack.